### PR TITLE
Refactors top-level module definitions.

### PR DESCRIPTION
**Pull request by yorkedork**
In addition to allowing feature selection, this change will help avoid some
potential (possible, not necessarily likely) import cycles.

### CODE REVIEW

**Comment by mdj2** [comment]
Making things that configurable isn't going to benefit me or the students since all the import time monkey patching is useful on our projects. However, if there is an actual *need* for opting out of certain features, then this is obviously a necessary change.

I'll give it a +0 if no one actually needs to opt out of features, otherwise a +1. @wylee what say you?

**Comment by wylee** [comment]
I'm +100 on this. Import-time monkey-patching makes me nervous and I don't like not being able to opt out of features. I only gave the actual changes a cursory glance; what I'm saying here is that I'm strongly in favor of moving related bits into modules and making things more explicit and configurable.

**Comment by yorkedork** [comment]
TLDR;

I initially had it "opt-in"; however, in the interests of maintaining backwards compatibility, it's an "opt-out" policy. I'm confused, though, @mdj2. I'm interpreting your statement to be "I'm +0 if ... otherwise a -1" which is considerably different from "I'm +0 if ... otherwise a +1" because context.

If you want to opt-out, you set a new attribute 'ARCUTILS_FEATURES', which is a dict that describes what you want or don't want.

For my cases (ORO), we either already perform some of the same run-time patching or don't want it. In order to avoid chaos and a-probably-too-unwieldy and large __init__ module, I decided to allow conditional feature flipping.

The initial observation was that all these bits shouldn't be all crammed into the __init__ module: all a consumer would need to do would be to import 'arcutils' or add it to installed apps and suddenly things are redefined. I consider this "Ruby behavior" and undesirable.

Additionally, I encountered import cycles, so the other motivation was to protect a bit more from that (as dependencies were imported at module scope in the __init__ module.)

**Comment by mdj2** [comment]
To simplify my stance on this PR: If ORO wants to use arcutils, +1, otherwise +0.

The motivation for creating arcutils was that students and I could drop it into our projects, and it does the right thing (according to me). Import-time side effects are frowned upon, but I consider a specialty package like this to be an appropriate use of them.

However, if ORO is buying into arcutils (which is not something I ever envisioned) then this feature is a must.

**Comment by yorkedork** [comment]
Oh! Ok. I understand your position now. This change wouldn't alter the behavior or intended purpose. It would only allow consumers to opt-out of some of it if they want.

I think it'd be valuable to have an ARC project utility package that will be broadly applicable: a package to provide DRYness to internal projects sharing the same environment(s). I like the idea of having such a package to provide a default logging configuration, among other things (off the top of my head) like deployment/release management tools, any sort of ES layer we might all decide we like, &c. To be honest, I was just on autopilot and wanted to configure ORO to use stashward. These are changes I made because I instinctively thought "add support for logstash forwarding, must install configure arcutils" and ran into a few hurdles looking through the code.

If we're bought into this vision, I see it as an advantage to modularize things a bit so we can toggle all the bits should the need arise. While I would avoid the kind of patch-up behavior we're seeing, I'm not being dismissive. What I meant earlier was that I'll pretty much always be +0 on that kind of code as long as it has an on/off switch. By putting all those things in `__init__` you're forcing anyone who imports the package to agree to the changes. By putting them in separate modules and having conditional logic (wherever you'd like that to be) seems to be a better approach because the default behavior can be (and is, in my changes) to perform all these fix-ups.

My only other thought is to avoid _activating_ all the logic in `__init__` by having the activation hooked into, say:

https://docs.djangoproject.com/en/1.7/ref/applications/#django.apps.AppConfig.ready

This would allow the 'give me everything by default' behavior to be toggle-able based on its presence in `INSTALLED_APPS`.

**Comment by wylee** [comment]
+1 to using AppConfig.ready().

**Comment by mdj2** [comment]
So it sounds we want arcutils to be *the* package that all Django projects, including the venerable ORO. Cool.

ORO isn't currently running Django 1.7, so AppConfig.ready() isn't available. I'm in favor of using that for the reason you stated. Is ORO planning to join the Python 3 club anytime soon (or at least Club 2.7)?

**Comment by mdj2** [comment]
Since ORO doesn't want any of the default features, we *could* just mock up `django.apps.AppConfig` if it can't be imported.

**Comment by yorkedork** [comment]
It is (I'd like to make progress towards 2&3'ing it during our next sprint which I hope will provide me the time); however, I don't know that it'd be a good approach to have us as a blocker.

What I might propose:

1. Perform any immediate cleanup using this pull request to provide 'initial support' in a new branch it (some branch `release/X`) and tag it.
2. Make additional changes to hook into `AppConfig.ready` in a distinct branch (some branch `release/Y`), and tag it.

I'll have ORO use the first release (1) until such time as it's ready to be welcomed (with tired, impatient arms) to the >1.7, >2.7 club.

**Comment by yorkedork** [comment]
I'd also support mocking `AppConfig` via import error capture.
